src/Gui/ShortcutManager.cpp

```diff
@@ -0,0 +1,455 @@
+    if (current == 0)
+        current = (int)actions.size();
+    setPriority(actions.front(), current);
+    ++current;
+    for (const auto &name : actions) {
+        int p = getPriority(name);
+        if (p <= 0 || p >= current) {
+            if (--current == 0)
+                --current;
+            setPriority(name, current);
+        } else
+            current = p;
+    }
+}
+
+int ShortcutManager::getPriority(const char *cmdName)
+{
+    if (!cmdName)
+        return 0;
+    auto it = priorities.find(cmdName);
+    if (it == priorities.end())
+        return 0;
+    return it->second;
+}
+
+void ShortcutManager::setPriority(const char *cmdName, int p)
+{
+    if (p == 0)
+        hPriorities->RemoveInt(cmdName);
+    else
+        hPriorities->SetInt(cmdName, p);
+}
+
+void ShortcutManager::setTopPriority(const char *cmdName)
+{
+    ++topPriority;
+    hPriorities->SetInt(cmdName, topPriority);
+}
+
+void ShortcutManager::onTimer()
+{
+    QAction *found = nullptr;
+    int priority = -INT_MAX;
+    int seq_length = 0;
+    for (const auto &info : pendingActions) {
+        if (info.action) {
+            info.action->setEnabled(true);
+            if (info.seq_length > seq_length
+                    || (info.seq_length == seq_length
+                        && info.priority > priority))
+            {
+                priority = info.priority;
+                seq_length = info.seq_length;
+                found = info.action;
+            }
+        }
+    }
+    if (found)
+        found->activate(QAction::Trigger);
+    pendingActions.clear();
+
+    if (lastFocus && lastFocus == QApplication::focusWidget()) {
+        // We are here because we have withheld some previous triggered action.
+        // We then disabled the action, and faked the same key strokes in order
+        // to wait for more for potential match of longer key sequence.  We use
+        // a timer to end the wait and triggered the pending action.
+        //
+        // However, Qt's internal shorcutmap state machine is still armed with
+        // our fake key strokes. So we try to fake some more obscure symbol key
+        // stroke below, hoping to reset Qt's state machine.
+
+        const auto &index = actionMap.get<1>();
+        static const std::string symbols = "~!@#$%^&*()_+";
+        QString shortcut = pendingSequence.toString() + QStringLiteral(", Ctrl+");
+        for (int s : symbols) {
+            QKeySequence k(shortcut + QLatin1Char(s));
+            auto it = index.lower_bound(ActionKey(k));
+            if (it->key.shortcut != k) {
+                QKeyEvent *kev = new QKeyEvent(QEvent::KeyPress, s, Qt::ControlModifier, 0, 0, 0);
+                QApplication::postEvent(lastFocus, kev);
+                kev = new QKeyEvent(QEvent::KeyRelease, s, Qt::ControlModifier, 0, 0, 0);
+                QApplication::postEvent(lastFocus, kev);
+                break;
+            }
+        }
+    }
+    timer.stop();
```

Comment:
It seems like this would make more sense at the top of the function, rather than the bottom.